exec.7
.TH "Miscellaneous" 9 "struct process_info" "August 2014" "API Manual" LINUX
.SH NAME
struct process_info \- process information
.SH SYNOPSIS
struct process_info {
.br
.BI "    pid_t " pi_pid ""
;

.br
.BI "    int " pi_stdin ""
;

.br
.BI "    int " pi_stdout ""
;

.br
.BI "    int " pi_stderr ""
;

.br
.BI "    int " pi_retval ""
;

.br
};
.br
.SH Members
.IP "pi_pid" 12
PID of the process
.IP "pi_stdin" 12
file descriptor connected to stdin of process pi_pid
.IP "pi_stdout" 12
file descriptor connected to stdout of process pi_pid
.IP "pi_stderr" 12
file descriptor connected to stderr of process pi_pid
.IP "pi_retval" 12
holds the exit status once the process has exited
.SH "Description"
This struct will be filled by \fBexec_process\fP to maintain
two-way communication with the child process once the function
returns. The file descriptors referring to stdin, stdout, and stderr
respectively of the child process have to be closed manually or by calling
\fBwait_for_child\fP with \fIclose_fds\fP set to true once they are of no use anymore
.TH "Miscellaneous" 9 "enum user_info_type" "August 2014" "API Manual" LINUX
.SH NAME
enum user_info_type \- user information type information
.SH SYNOPSIS
enum user_info_type {
.br
.BI "    USERINFO_TYPE_NONE"
, 
.br
.br
.BI "    USERINFO_TYPE_UID"
, 
.br
.br
.BI "    USERINFO_TYPE_NAME"

};
.SH Constants
.IP "USERINFO_TYPE_NONE" 12
invalid
.IP "USERINFO_TYPE_UID" 12
the provided information is a user's UID
.IP "USERINFO_TYPE_NAME" 12
the provided information is a user's name
.TH "exec_process" 9 "exec_process" "August 2014" "Kernel Hacker's Manual" LINUX
.SH NAME
exec_process \- execute a file
.SH SYNOPSIS
.B "_sentinel int" exec_process
.BI "(struct process_info *" proc ","
.BI "bool " wait ","
.BI "user_info_t " user ","
.BI "enum user_info_type " user_type ","
.BI "const char *" cmd ","
.BI "" ... ");"
.SH ARGUMENTS
.IP "proc" 12
storage space for process information
.IP "wait" 12
if set, wait until \fIcmd\fP terminates
.IP "user" 12
if set, drop privileges before executing \fIcmd\fP
.IP "user_type" 12
if \fIuser\fP is non-null, indicates the type of
user information (uid or name)
.IP "cmd" 12
the file to be executed
@...:			NULL-terminated list of arguments passed to \fIcmd\fP
.IP "..." 12
variable arguments
.SH "DESCRIPTION"
\fBexec_process\fP performs a \fBfork\fP and depending on its parameters, behaves as
.SH "FOLLOWS"
- if \fIwait\fP is set to true, \fIproc\fP is ignored and the function
returns only when \fIcmd\fP has terminated
- if \fIwait\fP is set to false, the function returns as soon as
\fIcmd\fP has been executed. Additionally, if \fIproc\fP is non-null,
it will be filled with the child's PID and file descriptors
to its standard input, output, and standard error
- if \fIuser\fP is non-null, the child drops its privileges after
forking. Supported types for the value specified in \fIuser\fP
are the user's UID or name. \fIuser_type\fP indicates which one
is actually used.
.TH "exec_process_p" 9 "exec_process_p" "August 2014" "Kernel Hacker's Manual" LINUX
.SH NAME
exec_process_p \- execute a file
.SH SYNOPSIS
.B "int" exec_process_p
.BI "(struct process_info *" proc ","
.BI "bool " wait ","
.BI "user_info_t " user ","
.BI "enum user_info_type " user_type ","
.BI "const char *" cmd ","
.BI "char *const " argv[] ");"
.SH ARGUMENTS
.IP "proc" 12
storage space for process information
.IP "wait" 12
if set, wait until \fIcmd\fP terminates
.IP "user" 12
if set, drop privileges before executing \fIcmd\fP
.IP "user_type" 12
if \fIuser\fP is non-null, indicates the type of
user information (uid or name)
.IP "cmd" 12
the file to be executed
.IP "argv[]" 12
NULL-terminated list of arguments passed to \fIcmd\fP
.SH "DESCRIPTION"
\fBexec_process\fP performs a \fBfork\fP and depending on its parameters, behaves as
.SH "FOLLOWS"
- if \fIwait\fP is set to true, \fIproc\fP is ignored and the function
returns only when \fIcmd\fP has terminated
- if \fIwait\fP is set to false, the function returns as soon as
\fIcmd\fP has been executed. Additionally, if \fIproc\fP is non-null,
it will be filled with the child's PID and file descriptors
to its standard input, output, and standard error
- if \fIuser\fP is non-null, the child drops its privileges after
forking. Supported types for the value specified in \fIuser\fP
are the user's UID or name. \fIuser_type\fP indicates which one
is actually used.
.TH "wait_for_child" 9 "wait_for_child" "August 2014" "Kernel Hacker's Manual" LINUX
.SH NAME
wait_for_child \- wait for a child process to terminate
.SH SYNOPSIS
.B "int" wait_for_child
.BI "(struct process_info *" proc ","
.BI "bool " close_fds ");"
.SH ARGUMENTS
.IP "proc" 12
process information
.IP "close_fds" 12
if true, open file descriptors to the child's
standard input, output and standard error are
closed upon process termination.
.TH "timed_read" 9 "timed_read" "August 2014" "Kernel Hacker's Manual" LINUX
.SH NAME
timed_read \- read from a file descriptor
.SH SYNOPSIS
.B "ssize_t" timed_read
.BI "(int " fd ","
.BI "void *" buf ","
.BI "size_t " size ","
.BI "unsigned int " timeout ");"
.SH ARGUMENTS
.IP "fd" 12
file descriptor to read from
.IP "buf" 12
buffer for the returned data
.IP "size" 12
size of the buffer pointed to by \fIbuf\fP
.IP "timeout" 12
time to wait for data
.SH "DESCRIPTION"
\fBtimed_read\fP attempts to read up to \fIsize\fP bytes from file descriptor \fIfd\fP
into the buffer starting at \fIbuf\fP. If \fItimeout\fP was set to 0, this
function behaves exactly like read(2). Otherwise, select(2) will be used
to wait up to \fItimeout\fP seconds until returning an error.
.TH "timed_write" 9 "timed_write" "August 2014" "Kernel Hacker's Manual" LINUX
.SH NAME
timed_write \- write to a file descriptor
.SH SYNOPSIS
.B "ssize_t" timed_write
.BI "(int " fd ","
.BI "const void *" buf ","
.BI "size_t " size ","
.BI "unsigned int " timeout ");"
.SH ARGUMENTS
.IP "fd" 12
file descriptor to write to
.IP "buf" 12
buffer containing the data to be written
.IP "size" 12
size of the buffer pointed to by \fIbuf\fP
.IP "timeout" 12
time to wait for \fIfd\fP to become ready
.SH "DESCRIPTION"
\fBtimed_write\fP attempts to write up to \fIsize\fP bytes from the buffer starting
at \fIbuf\fP to the file descriptor \fIfd\fP. If \fItimeout\fP was set to 0, this
function behaves exactly like write(2). Otherwise, select(2) will be used
to wait up to \fItimeout\fP seconds until returning an error.
.TH "get_exit_details" 9 "get_exit_details" "August 2014" "Kernel Hacker's Manual" LINUX
.SH NAME
get_exit_details \- get information about a process exit value
.SH SYNOPSIS
.B "void" get_exit_details
.BI "(int " status ","
.BI "int *" ret ","
.BI "bool *" core ","
.BI "bool *" signaled ","
.BI "bool *" parent ");"
.SH ARGUMENTS
.IP "status" 12
exit status
.IP "ret" 12
the real return value
.IP "core" 12
indicator if core was dumped
.IP "signaled" 12
indicator if process was killed by a signal
.IP "parent" 12
indicator if error has occurred in the parent
process (only valid if \fIstatus\fP < 0)
.SH "DESCRIPTION"
\fBget_exit_details\fP takes the process exit value \fIstatus\fP (typically
obtained by \fBwaitpid\fP and co) and sets the parameters as follows:
- if the process exited normally (exit, _exit, or return from main),
\fIret\fP will be set to the exit value
- if the process was killed by a signal, \fIsignaled\fP will be set to
true and \fIret\fP will be set to the signal that caused the abortion
- if the process was killed by a signal, \fIcore\fP will indicate if a
core dump has been created when the process died (and as before,
\fIret\fP will hold the signal number)
.TH "copy_exit_detail_str" 9 "copy_exit_detail_str" "August 2014" "Kernel Hacker's Manual" LINUX
.SH NAME
copy_exit_detail_str \- obtain a printable description for an exit status
.SH SYNOPSIS
.B "int" copy_exit_detail_str
.BI "(int " status ","
.BI "char **" buffer ");"
.SH ARGUMENTS
.IP "status" 12
exit status
.IP "buffer" 12
pointer to a to-be-allocated buffer
.SH "DESCRIPTION"
\fBcopy_exit_detail_str\fP takes the process exit value \fIstatus\fP (typically
obtained by \fBwaitpid\fP and co) and fills the buffer \fIbuffer\fP with a
printable description. Information includes:
- the return code
- if killed by a signal, the signal number
- if killed by a signal, whether the core was dumped


exec.c
/*
 * =============================================================================
 *
 *       Filename:  exec.c
 *
 *    Description:  Collection of function to spawn and interact
 *                  with child processes
 *
 *        Version:  1.0
 *        Created:  08/21/2014 07:45:49 PM
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Oliver Lorenz (ol), olli@olorenz.org
 *
 * =============================================================================
 */

#include <errno.h>
#include <fcntl.h>
#include <grp.h>
#include <pwd.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include "exec.h"

#define ROOT_UID		0

#define NUM_PIPES		3

#define PIPE_STDIN		0
#define PIPE_STDOUT		1
#define PIPE_STDERR		2

#define PIPE_RD_FD		0
#define PIPE_WR_FD		1

static int drop_privileges(const struct passwd *user)
{
	if (setgid(user->pw_gid))
		return 1;

	if (initgroups(user->pw_name, user->pw_gid))
		return 1;

	if (setuid(user->pw_uid))
		return 1;

	if (setuid(ROOT_UID) == 0) {
		errno = EPERM;
		return 1;
	}

	return 0;
}

static struct passwd *resolve_user(user_info_t user, enum user_info_type type)
{
	struct passwd *_user = NULL;

	errno = 0;
	switch (type) {
	case USERINFO_TYPE_UID:
		_user = getpwuid(*(user.ui_uid));
		break;
	case USERINFO_TYPE_NAME:
		_user = getpwnam(user.ui_name);
		break;
	case USERINFO_TYPE_NONE:
		/* fallthrough */
	default:
		errno = EINVAL;
		break;
	}

	if (!_user && !errno)
		errno = ENOENT;

	return _user;
}

int wait_for_child(struct process_info *proc, bool close_fds)
{
	pid_t child;

	do {
		child = waitpid(proc->pi_pid, &(proc)->pi_retval, 0);
	} while (child == (pid_t) - 1 && errno == EINTR);

	if (close_fds) {
		close(proc->pi_stdin);
		close(proc->pi_stdout);
		close(proc->pi_stderr);
		proc->pi_stdin  = -1;
		proc->pi_stdout = -1;
		proc->pi_stderr = -1;
	}

	return (child == (pid_t) - 1 ? -1 : 0);
}

void get_exit_details(int status, int *ret, bool *core, bool *signaled, bool *parent)
{
	(*ret)      = -1;
	(*core)     = false;
	(*signaled) = false;
	(*parent)   = false;

	if (status < 0) {
		if (status > -EXEC_PROCESS_ERROR_OFFSET) {
			/* status contains parent's errno */
			errno = -status;
			(*parent) = true;
		} else {
			status += EXEC_PROCESS_ERROR_OFFSET;
			errno = -status;
		}
		return;
	}

	if (unlikely(WIFSIGNALED(status))) {
		if (WCOREDUMP(status))
			(*core) = true;
		(*signaled) = true;
		(*ret) = WTERMSIG(status);
	} else if (WIFEXITED(status)) {
		(*ret) = WEXITSTATUS(status);
	}
}

int copy_exit_detail_str(int status, char **buffer)
{
	int ret;
	bool core;
	bool signaled;
	bool parent;

	get_exit_details(status, &ret, &core, &signaled, &parent);
	if (ret == -1) {
		if (parent)
			ret = asprintf(buffer, "An error has occurred in the "
			               "parent process: %s", strerror(errno));
		else
			ret = asprintf(buffer, "An error has occurred in the "
			               "child process: %s", strerror(errno));
	} else {
		if (signaled)
			ret = asprintf(buffer, "Process exited due to signal "
			               "%d (core %s dumped)", ret,
			               core ? "was" : "not");
		else
			ret = asprintf(buffer, "Process exited with status %d",
			               ret);
	}

	return (ret == -1);
}

static inline int fd_get_flags(int fd)
{
	return fcntl(fd, F_GETFL, NULL);
}

static inline int fd_set_flags(int fd, int flags)
{
	return fcntl(fd, F_SETFL, flags);
}

static inline int fd_set_nonblocking(int fd, int flags)
{
	return fd_set_flags(fd, flags | O_NONBLOCK);
}

static inline int fd_clear_nonblocking(int fd, int flags)
{
	return fd_set_flags(fd, flags & (~O_NONBLOCK));
}

static inline int fd_make_nonblocking(int fd)
{
	int flags;
	flags = fd_get_flags(fd);
	if (flags == -1)
		return -1;
	return fd_set_nonblocking(fd, flags);
}

ssize_t timed_read(int fd, void *buf, size_t size, unsigned int timeout)
{
	int ret;
	int flags;
	fd_set set;
	size_t have;
	ssize_t count;
	struct timeval wait_time;

	flags = fd_get_flags(fd);
	if (flags == -1)
		return -1;

	if (fd_set_nonblocking(fd, flags))
		return -1;

	if (!timeout)
		timeout = (unsigned int) -1;

	have = 0;
	for (;;) {
		FD_ZERO(&set);
		FD_SET(fd, &set);

		wait_time.tv_sec = timeout;
		wait_time.tv_usec = 0;

		ret = select(fd + 1, &set, NULL, NULL, &wait_time);
		if (ret == 0) {
			if (!have) {
				errno = ETIMEDOUT;
				have = -1;
			}
			break;
		} else if (ret == -1) {
			if (errno == EINTR) {
				continue;
			} else {
				have = -1;
				break;
			}
		} else {
			count = read(fd, buf + have, size - have);
			if (count == -1) {
				if (errno == EINTR || errno == EAGAIN) {
					continue;
				} else {
					have = -1;
					break;
				}
			} else if (count == 0) {
				break;
			} else {
				have += (size_t) count;
			}
		}
	}

	(void) fd_clear_nonblocking(fd, flags);
	return have;
}

ssize_t timed_write(int fd, const void *buf, size_t size, unsigned int timeout)
{
	int ret;
	int flags;
	fd_set set;
	size_t have;
	ssize_t count;
	struct timeval wait_time;

	flags = fd_get_flags(fd);
	if (flags == -1)
		return -1;

	if (fd_set_nonblocking(fd, flags))
		return -1;

	if (!timeout)
		timeout = (unsigned int) -1;

	have = 0;
	for (;;) {
		FD_ZERO(&set);
		FD_SET(fd, &set);

		wait_time.tv_sec = timeout;
		wait_time.tv_usec = 0;

		ret = select(fd + 1, NULL, &set, NULL, &wait_time);
		if (ret == 0) {
			if (!have) {
				errno = ETIMEDOUT;
				have = -1;
			}
			break;
		} else if (ret == -1) {
			if (errno == EINTR) {
				continue;
			} else {
				have = -1;
				break;
			}
		} else {
			count = write(fd, buf + have, size - have);
			if (count == -1) {
				if (errno == EINTR || errno == EAGAIN) {
					continue;
				} else {
					have = -1;
					break;
				}
			} else if (count == 0) {
				break;
			} else {
				have += (size_t) count;
			}
		}
	}

	(void) fd_clear_nonblocking(fd, flags);
	return have;
}

_sentinel int exec_process(struct process_info *proc_info, bool wait,
                           user_info_t user, enum user_info_type user_type,
                           const char *cmd, ...)
{
	int res;
	unsigned int i, num_args;
	char **args = NULL;
	va_list ap;

	num_args = 0;
	va_start(ap, cmd);
	while (va_arg(ap, char *))
		++num_args;
	va_end(ap);

	/* cmd + NULL */
	num_args += 2;

	if ((args = malloc(num_args * sizeof(*args))) == NULL) {
		res = -errno;
		goto exit;
	}

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wcast-qual"
	args[0] = (char *)cmd;
#pragma GCC diagnostic pop

	va_start(ap, cmd);
	for (i = 0; i < num_args - 2; ++i)
		args[i + 1] = va_arg(ap, char *);
	va_end(ap);

	args[num_args - 1] = NULL;

	res = exec_process_p(proc_info, wait, user, user_type, cmd, args);
exit:
	free(args);
	return res;
}

int exec_process_p(struct process_info *proc_info, bool wait,
                   user_info_t user, enum user_info_type user_type,
                   const char *cmd, char *const argv[])
{
	int pipes[NUM_PIPES][2];
	int self_pipe[2] = { -1, -1 };
	unsigned int i;
	int flags, child_error;
	pid_t pid;
	int res;

	for (i = 0; i < NUM_PIPES; ++i) {
		pipes[i][PIPE_RD_FD] = -1;
		pipes[i][PIPE_WR_FD] = -1;
	}

	if (wait) {
		/* Not needed if we're waiting */
		proc_info = NULL;
	}

	if (proc_info) {
#ifdef __linux__
		if (pipe2(pipes[ PIPE_STDIN], 0) ||
		    pipe2(pipes[PIPE_STDOUT], O_NONBLOCK) ||
		    pipe2(pipes[PIPE_STDERR], O_NONBLOCK)) {
			res = -errno;
			goto exit;
		}
#else
		if (pipe(pipes[ PIPE_STDIN]) ||
		    pipe(pipes[PIPE_STDOUT]) ||
		    pipe(pipes[PIPE_STDERR])) {
			res = -errno;
			goto exit;
		}

		if (fd_make_nonblocking(pipes[PIPE_STDOUT][PIPE_RD_FD]) ||
		    fd_make_nonblocking(pipes[PIPE_STDOUT][PIPE_WR_FD]) ||
		    fd_make_nonblocking(pipes[PIPE_STDERR][PIPE_RD_FD]) ||
		    fd_make_nonblocking(pipes[PIPE_STDERR][PIPE_WR_FD])) {
			res = -errno;
			goto exit;
		}
#endif
	}

	if (pipe(self_pipe)) {
		res = -errno;
		goto exit;
	}

	flags = fcntl(self_pipe[PIPE_WR_FD], F_GETFD);
	if (flags == -1) {
		res = -errno;
		goto exit;
	}

	if (fcntl(self_pipe[PIPE_WR_FD], F_SETFD, flags | FD_CLOEXEC)) {
		res = -errno;
		goto exit;
	}

	pid = fork();
	if (pid == 0) {
		/* child */
		long maxfd;
		struct passwd *_user = NULL;

		if (proc_info) {
			dup2(pipes[ PIPE_STDIN][PIPE_RD_FD], STDIN_FILENO);
			dup2(pipes[PIPE_STDOUT][PIPE_WR_FD], STDOUT_FILENO);
			dup2(pipes[PIPE_STDERR][PIPE_WR_FD], STDERR_FILENO);

			close(pipes[ PIPE_STDIN][PIPE_RD_FD]);
			close(pipes[PIPE_STDOUT][PIPE_WR_FD]);
			close(pipes[PIPE_STDERR][PIPE_WR_FD]);

			close(pipes[ PIPE_STDIN][PIPE_WR_FD]);
			close(pipes[PIPE_STDOUT][PIPE_RD_FD]);
			close(pipes[PIPE_STDERR][PIPE_RD_FD]);
		}

		close(self_pipe[PIPE_RD_FD]);

		_user = resolve_user(user, user_type);
		if (!_user && errno != EINVAL /* USERINFO_TYPE_NONE */ )
			goto fail;

		if (_user && drop_privileges(_user))
			goto fail;

		maxfd = sysconf(_SC_OPEN_MAX);
		while (--maxfd > 3) {
			if (maxfd == self_pipe[PIPE_WR_FD])
				continue;
			close((int)maxfd);
		}

		execvp(cmd, argv);

fail:
		child_error = EXEC_PROCESS_ERROR_OFFSET + errno;
		write(self_pipe[PIPE_WR_FD], &child_error, sizeof(child_error));
		_exit(1);
	} else if (pid > 0) {
		/* parent */
		ssize_t count;

		if (proc_info) {
			close(pipes[ PIPE_STDIN][PIPE_RD_FD]);
			close(pipes[PIPE_STDOUT][PIPE_WR_FD]);
			close(pipes[PIPE_STDERR][PIPE_WR_FD]);
		}

		close(self_pipe[PIPE_WR_FD]);

		while ((count =
			read(self_pipe[PIPE_RD_FD], &child_error,
			     sizeof(child_error))) == -1) {
			if (errno != EAGAIN && errno != EINTR)
				break;
		}

		close(self_pipe[PIPE_RD_FD]);

		if (count) {
			res = -child_error;
			goto exit;
		}

		if (wait) {
			int ret;
			pid_t child;

			do {
				child = waitpid(pid, &ret, 0);
			} while (child == (pid_t) -1 && errno == EINTR);

			res = (child == (pid_t) -1 ? -errno : ret);
		} else {
			res = 0;
		}

		if (proc_info) {
			proc_info->pi_pid = pid;
			proc_info->pi_stdin  = pipes[ PIPE_STDIN][PIPE_WR_FD];
			proc_info->pi_stdout = pipes[PIPE_STDOUT][PIPE_RD_FD];
			proc_info->pi_stderr = pipes[PIPE_STDERR][PIPE_RD_FD];
		}
	} else {
		res = -errno;
		goto exit;
	}

	return res;

exit:
	(void) close(self_pipe[PIPE_RD_FD]);
	(void) close(self_pipe[PIPE_WR_FD]);
	for (i = 0; i < NUM_PIPES; ++i) {
		(void) close(pipes[i][PIPE_RD_FD]);
		(void) close(pipes[i][PIPE_WR_FD]);
	}
	if (proc_info)
		memset(proc_info, 0, sizeof(*proc_info));
	return res;
}
